### DISCUSSION TITLE
vcpkg overlay-port command seems not been considered when configuring a dependency with vcpkg_configure_cmake

### DISCUSSION THREAD

**Discussion by andriy-fedechko-imprivata** [microsoft/vcpkg#37194] [Q&A] [answered]
Hello, 

I faced with the case that I can't figure out. That's might be something related to my lack of knowledge or something that might actually be a real case. 

My problem is next:
I have a **project A** that has one of it's dependency as an _overlay-port_ (not sure if that's right term to use, though). 
Basically, in **project's A** root folder I have _vcpkg-ports_ folder that then contains another folder for **project B** with related _portfile.cmake_ and _vcpkg.json_ files. 

In _portfile.cmake_ I have next code to configure **project B**:

```
// get stuff from git
vcpkg_from_github(....)

// configuring the project
vcpkg_configure_cmake(
    SOURCE_PATH
    ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DBUILD_TESTING=OFF
        -DVCPKG_OVERLAY_PORTS="${SOURCE_PATH}/vcpkg-ports"
    )

// do install and other things
vcpkg_install_cmake()
vcpkg_fixup_cmake_targets(...)
vcpkg_copy_pdbs()

```

I have **project B** listed within **project's A** _vcpkg.json_ file. So that vcpkg knows that **project A** depends on **project B**.

In **project's A** _CMakePresets_ file i have defined a _"VCPKG_OVERLAY_PORTS": "${sourceDir}/vcpkg-ports"_ so that once I configure **project A** vcpkg knows where to look for **project B**. Basically look in the _vcpkg-ports_ folder 

So far so good. Once I start configure **project A** vcpkg correctly detects that it depends on **project B** and it also detects **project's B** data in _vcpkg-ports_ folder and tries to configure project B first. 

And here is where problem happens. 
**Project B** has the same configuration, the project depends on something from it's own _vcpkg-port_ folder, e.g. on **project C**. 
So while configuring **project B** as part of **project's A** configuration the configuration fails and the log says that it can't find **project's B** port that **project B** has in it's _vcpkg-ports_ folder. Basically it can't find **project C**

However, if I configure **project B** manually, I mean not as  dependency for some other project but a a standalone project, **project's B** dependency is correctly detected, configured and built. 

Would appreciate any help with that. 

Thanks in advance.

**Comment by Neumann-A** [comment, marked as answer]
```
// configuring the project
vcpkg_configure_cmake(
    SOURCE_PATH
    ${SOURCE_PATH}
    PREFER_NINJA
    OPTIONS
        -DBUILD_TESTING=OFF
        -DVCPKG_OVERLAY_PORTS="${SOURCE_PATH}/vcpkg-ports"
    )
```

That is not how vcpkg works. 

Are you trying to use vcpkg within vcpkg? That is probably not supported.

**Reply by andriy-fedechko-imprivata** [reply]
looks like. 

so basically, when I do a call: vcpkg_configure_cmake it will not force vcpkg to iterate over that port's dependencies? All the dependencies for that particular project (in my case **project B**) should be mentioned in _vcpkg.json_ file in **project's A** overlay port?  

At this point **project A** must list all project's B dependencies and must take them before configuring **project B**?

**Reply by Neumann-A** [reply]
You would simply make `portA` and `portB` and list the deps accordingly. To provide the overlays you would combine them to
`-DVCPKG_OVERLAY_PORTS="overlayA;overlayB"` in project C so that port A and B gets build for project C.

**Reply by andriy-fedechko-imprivata** [reply]
okay, so from your example, basically what it means is that even that projectC (in my case) depends only on `portA` and `portA` depends on `portB` I still need to list both for my **project's  C** dependencies?

**Reply by Neumann-A** [reply]
projects C manifest can only depend on port A since port A will pull in port B by its manifest. However you need to provide a consistent registry for both so the overlay needs to know both.

**Reply by andriy-fedechko-imprivata** [reply]
yeah... 
so I do have a registry, e.g. my file vcpkg-configuration.json
```
{
  "default-registry": {
    "kind": "git",
    "repository": "https://github.com/my-project/custom-vcpkg-registry",
    "baseline": "5568f110b509a9fd90711978a7cb76bae75bb092"
  }
}
```
that _custom-vcpkg-registry_ contains all ports that all my projects depends on. So if my project depends of port from git registry - all if fine. 

but here is different, my **project A** depends on `portB ` and that `portB` depends on `portC` and that `portC` is not in git registry.


I guess I'm trying to understand - how in my case all the dependencies that are in my _custom-vcpkg-registry_ are properly taken and configuring for dependent projects, but that one dependency that is defined locally can't be found. 

that's projcet A
![image](https://github.com/microsoft/vcpkg/assets/80824083/3ad9e2a3-8c81-4e41-9f58-146974579c26)

that's project B
![image](https://github.com/microsoft/vcpkg/assets/80824083/48e7c7aa-8856-444c-8a0e-c17cc05ab2fc)

eventually project B will be moved into _custom-vcpkg-registry_ so that projectA no longer depends on local port and doesn't have overlay-port defined, and takes it dependency from _custom-vcpkg-registry_ directly. 

but before doing that change I decided to try with overlay to have everything properly configured in projectB

**Reply by andriy-fedechko-imprivata** [reply]
@Neumann-A  thank you a lot with your attempts to gently explain that I'm doing things in a way that is unappropriated. After giving more thoughts on this I think I understood your points. 
the main one is basically that  `vcpkg_configure_cmake` is just forces _CMake_ to do it's stuff and doesn't do any _vcpkg_ related actions, such as download dependencies, etc. 

With that, I think I also have an idea how to solve my particular problem in proper way. 

Thank a lot.
